builtins/amp-img.js

```diff
@@ -239,7 +242,26 @@ export class AmpImg extends BaseElement {
     if (this.getLayoutWidth() <= 0) {
       return Promise.resolve();
     }
-    return this.loadPromise(img);
+    return this.loadPromise(img).then(() => this.warnAgainstOversizedImages_());
+  }
+
+  /**
+   * @return {!Promise}
+   * @private
+   */
+  warnAgainstOversizedImages_() {
+    return this.measureElement(() => {
+      const {naturalWidth, naturalHeight} = this.img_;
+      const {scrollWidth, scrollHeight} = this.element;
+      const widthRatio = naturalWidth / scrollWidth;
+      const heightRatio = naturalHeight / scrollHeight;
+      if (widthRatio > this.getDpr() * 2 || heightRatio > this.getDpr() * 2) {
```

Comment:
As things stand, this is currently only a user warning, not an error. 

We thought about optimizing this check for cases of `object-fit: contain / cover` as well as for `transform`, since in those cases the displayed image dimensions will be different from the client dimensions. However, that would be extra payload for relatively uncommon use cases, and as is, `this.user().warn` results in extra payload for a developer warning, which doesn't directly benefit the user. 

We're wondering if it's possible to ship a developer version of the code with robust and comprehensive errors to developers for testing / development, that's separate from our current minified bundle. I think logistically, we already have this, but practically, developers don't use it that way. 

What are your thoughts, @kristoferbaxter?

Reply:
The increase in bundle size for the current version is quite small, not great but makes sense **if** the value of the warning is quite high.

Otherwise, it might make more sense to implement this during validation or a separate development bundle since it doesn't break user experience.

Reply:
Okay, I'm going to let @jridgewell decide if we want to merge this current version as is (or with changes) for #302, or just... leave #302. =)